lib/roger/partition/worker.ex

```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
             GProc.regp({:roger_job_worker_meta, state.partition_id, job.id}, job)
-
-            before_run_state = callback(:before_run, [state.partition_id, job])
-            try do
-              result = Job.execute(job)
-              job_done(job, :ack, state)
-
-              callback(:after_run, [state.partition_id, job, result, before_run_state])
-            catch
-              t, e ->
-                #Logger.error "Execution error: #{t}:#{inspect e}"
-              cb = if Job.retryable?(job) do
-                case Retry.retry(state.channel, state.partition_id, job) do
-                  {:ok, :buried} -> :on_buried
-                  {:ok, _expiration} -> :on_error
-                end
-              else
-                :on_error
-              end
-
-              job_done(job, :ack, state)
-              callback(cb, [state.partition_id, job, {t, e}, System.stacktrace(), before_run_state])
-
-              GProc.unregp(name(job.id))
-              GProc.unregp({:roger_job_worker_meta, state.partition_id, job.id})
-
+            parent = self()
+            {pid, _ref} = spawn_monitor(fn () ->
+              execute_job(job, state, parent)
+              send(parent, :job_finished)
+            end)
+
+            if is_number(job.max_execution_time) do
+              Process.send_after(self(), :handle_job_timeout, job.max_execution_time)
             end
+
+            {:noreply, %{state | worker_task_pid: pid, job: job}}
           end
         end
       {:error, message} ->
         # Decode error
         Logger.debug "Job decoding error: #{inspect message} #{inspect state.raw_payload}"
         job_done(nil, :ack, state)
+        {:stop, :normal, state}
     end
+
+  end
+
+  def handle_info(:job_finished, state) do
+    {:stop, :normal, state}
+  end
+
+  def handle_info(:job_errored, state) do
+    GProc.unregp(name(state.job.id))
+    GProc.unregp({:roger_job_worker_meta, state.partition_id, state.job.id})
     {:stop, :normal, state}
   end
 
+  def handle_info({:DOWN, _ref, :process, _child, reason}, state) do
+    handle_error(state.job, {:worker_crash, reason}, nil, state)
+    {:stop, :normal, state}
+  end
+
+  def handle_call(:cancel_job, _source, state) do
+    Process.exit(state.worker_task_pid, :kill)
+    {:reply, :ok, state, 0}
+  end
+
+  defp execute_job(job, state, parent) do
+    before_run_state = callback(:before_run, [state.partition_id, job])
+    try do
+      result = Job.execute(job)
+
+      job_done(job, :ack, state)
+
+      callback(:after_run, [state.partition_id, job, result, before_run_state])
+    catch
+      type, exception ->
+        #Logger.error "Execution error: #{t}:#{inspect e}"
+        handle_error(job, {type, exception}, before_run_state, state)
+        send(parent, :job_errored)
+    end
+  end
+
+  defp handle_error(job, {type, exception}, before_run_state, state) do
+    cb = if Job.retryable?(job) do
+      case Retry.retry(state.channel, state.partition_id, job) do
```

Comment:
You can probably combine the if in the case statement as well. The last function executed in a case statement is return so you either get back a retry response, or `false`

Reply:
Or use a `with` expression